File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        charIndexMap ={}
        left = 0
        maxLength =0

        for right in range(len(s)):
            if s[right] in charIndexMap and charIndexMap[s[right]] >= left:
                left = charIndexMap[s[right]] + 1
            
            charIndexMap[s[right]] = right
            maxLength = max(maxLength, right - left +1)

        return maxLength        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndexMap;
        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            if (charIndexMap.find(s[right]) != charIndexMap.end()
            && charIndexMap[s[right]] >=left) {
                left = charIndexMap[s[right]] +1;
            }

            charIndexMap[s[right]] = right;
            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }
};